### PR TITLE
OCLOMRS-478: When adding a CIEL mapping or Bulk Concepts, execute a search only when ENTER is pressed

### PR DESCRIPTION
**Pull request by judeinno**
# JIRA TICKET NAME:
[When adding a CIEL mapping or Bulk Concepts, execute a search only when ENTER is pressed](https://issues.openmrs.org/browse/OCLOMRS-478)

# Summary:
When adding a CIEL mapping, execute a search only when ENTER is pressed in order to be consistent with other searches, and to avoid multiple calls to the backend at every keystroke. Repeat the same behaviour for the Add Bulk Concepts search fields.

### CODE REVIEW

File: src/components/dictionaryConcepts/components/CreateMapping.jsx [hunk 1]
```diff
@@ -8,19 +7,35 @@ import MapType from './MapType';
...
-    this.state = {};
+    this.state = { inputValue: this.props.to_concept_name || '', options: [], isVisible: false };
   }
 
   handleInputChange = (value) => {
     this.setState({ inputValue: value });
   }
 
+  handleKeyPress = (e, inputValue, url) => {
+    if (String(e.key).toUpperCase() === 'ENTER') {
```
Comment (dkayiwa) [inline src/components/dictionaryConcepts/components/CreateMapping.jsx]: Same as above

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 2]
```diff
@@ -114,6 +115,15 @@ export class AddBulkConcepts extends Component {
     this.setState({ openResultModal: false });
   }
 
+  onKeyDown = (event, onKeyDown, onChange) => {
+    this.setState({ value: event.target.value });
+    if (event.keyCode === 13) {
```
[the remark this answers is not in the document]
Comment (judeinno) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: Its the key code for when a user presses the enter button.

File: src/components/dictionaryConcepts/components/CreateMapping.jsx (hunk 1 above)
Comment (judeinno) [reply src/components/dictionaryConcepts/components/CreateMapping.jsx]: Yes at that point I convert the key entered to uppercase and check if its equivalent to `ENTER`

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkayiwa) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: Will you keep around to keep answering who ever will be asking what this 13 means?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 3]
```diff
@@ -74,4 +74,183 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+
+  it('should call on change function', () => {
```
[the remark this answers is not in the document]
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Let me update all of them.

[the remark this answers is not in the document]
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: @dkayiwa, I have updated the tests titles please review.

File: package.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
     "react-router-dom": "^4.2.2",
     "react-scripts": "1.1.4",
     "react-select": "^2.4.2",
+    "react-select-plus": "^1.0.0-rc.3.patch12",
```
Comment (dkayiwa) [inline package.json]: How is this related to the ticket?

Comment (judeinno) [reply package.json]: I have updated it was a library I was using but I removed it.

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 5]
```diff
@@ -114,6 +115,14 @@ export class AddBulkConcepts extends Component {
     this.setState({ openResultModal: false });
   }
 
+  onKeyDown = (event, onKeyDown, onChange) => {
+    if (String(e.key).toUpperCase() === 'ENTER') {
```
Comment (dkayiwa) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: What happens when you need to change from ENTER to ENTRE?

Comment (judeinno) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: Please clarify, I don't understand what you mean by `ENTRE`

Comment (dkayiwa) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: What happens if you discover that you misspelt ENTER?

File: src/components/dictionaryConcepts/components/CreateMapping.jsx (hunk 1 above)
Comment (danuluma) [reply src/components/dictionaryConcepts/components/CreateMapping.jsx]: We could listen for the keyCode instead,,, it's value for the enter key is usually 13

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (judeinno) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have updated this with `(String(e.key).toUpperCase() === 'ENTER')`

[the remark this answers is not in the document]
Comment (dkayiwa) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: Does that work for non English key boards?

File: src/components/dictionaryConcepts/components/CreateMapping.jsx (hunk 1 above)
Comment (dkayiwa) [reply src/components/dictionaryConcepts/components/CreateMapping.jsx]: Will this work for non English key boards?

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (judeinno) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have updated it by setting a constant `const KEY_CODE_FOR_ENTER = 13;` that I use to Identify the key code and also gave it a descriptive name for a new user

File: src/components/dictionaryConcepts/components/CreateMapping.jsx (hunk 1 above)
Comment (judeinno) [reply src/components/dictionaryConcepts/components/CreateMapping.jsx]: I have updated it by setting a constant `const KEY_CODE_FOR_ENTER = 13;` that I use to Identify the key code and also gave it a descriptive name for a new user

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 5 above)
Comment (judeinno) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have updated it by setting a constant `const KEY_CODE_FOR_ENTER = 13;` that I use to Identify the key code and also gave it a descriptive name for a new user

**Comment by dkayiwa** [comment]
Did you look at the merge conflicts?

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx [hunk 6]
```diff
@@ -113,14 +113,15 @@ describe('Add Bulk Concepts', () => {
       </Provider>
     </MemoryRouter>);
     const input = wrapper.find('#search');
-    input.simulate('change', { target: { value: '' } });
+    input.simulate('change', { target: { value: 'Bronze Diabetes' } });
     input.simulate('keydown', { key: 'ArrowDown' });
-    input.simulate('keydown', { key: 'Enter' });
+    input.simulate('keydown', { keyCode: 13 });
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/addBulkConcepts.test.jsx]: Didn't you create a constant for the above?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 7]
```diff
@@ -74,4 +74,183 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+
+  it('should test update new mappings for new rows', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: What are you testing about new mappings for new rows?

Comment (dkayiwa) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Secondly, this being a test, aren't do we need to include the word test?

Comment (dkayiwa) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: The same applies everywhere below

Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated the test descriptions please review.

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx (hunk 6 above)
Comment (judeinno) [reply src/tests/bulkConcepts/addBulkConcepts.test.jsx]: I have updated this to use the constant I made for the key code.

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 8]
```diff
@@ -74,4 +75,183 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+
+  it('should simulate update of new mappings row data', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Is the page being tested supposed to simulate update of nee mappings row data as part of its functionality?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 9]
```diff
@@ -74,4 +75,183 @@ describe('Test suite for dictionary concepts components', () => {
...
+
+    wrapper = mount(<Router>
+      <table><tbody><CreateMapping {...newProps} /></tbody></table>
+    </Router>);
+    const inputField = wrapper.find('CreateMapping');
+    inputField.find('#source').simulate('change');
+    expect(props.updateEventListener).toHaveBeenCalled();
+  });
+
+  it('should simulate update of new mappings for non Ciel concepts', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Same as above

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 8 above)
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated it, please review.

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 9 above)
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated it, please review.

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 10]
```diff
@@ -74,4 +75,183 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+
+  it('should update of new mappings row data', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: In the above description, when should the update of new mappings of row data take place?
The same applies every where below.

Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Please review again and see if its a more suitable description.

**Comment by dkayiwa** [comment]
Did you see the merge conflicts?

File: src/styles/dictionary_concepts.scss [hunk 11]
```diff
@@ -411,3 +411,31 @@ input:-webkit-autofill:active {
 input[type=text] {
   text-overflow: ellipsis;
 }
+
+.conceptDetails {
+  max-height: 80px;
+}
+.cielConceptsList {
+  border: 1px solid #bbb;
```
Comment (dkayiwa) [inline src/styles/dictionary_concepts.scss]: How is this related to execute a search only when ENTER is pressed?

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx [hunk 12]
```diff
@@ -214,7 +216,7 @@ describe('Add Bulk Concepts', () => {
     const other = wrapper.find('#otherSourcesOption');
     other.simulate('click');
     const input = wrapper.find('#sourceSearch');
-    input.simulate('change', { target: { value: '' } });
+    input.simulate('change', { target: { value: 'testSource' } });
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/addBulkConcepts.test.jsx]: How is this related to execute a search only when ENTER is pressed?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 13]
```diff
@@ -86,4 +87,183 @@ describe('Test suite for dictionary concepts components', () => {
     wrapper.setProps(nextProps);
     expect(wrapper.instance().props.map_type).toEqual('Associated With');
   });
+
+  it('should handle update of new mappings row source data when isNew is true', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: How are these tests related to execute a search only when ENTER is pressed?

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx (hunk 12 above)
Comment (judeinno) [reply src/tests/bulkConcepts/addBulkConcepts.test.jsx]: Because to do a search one is supposed to first press enter hence I first had to simulate change with the value I would like to search for.

File: src/styles/dictionary_concepts.scss (hunk 11 above)
Comment (judeinno) [reply src/styles/dictionary_concepts.scss]: Since the library(AsyncSelect) that was being used could not allow one to press enter to search I had to recreate my own code that creates a dropdown and allows selecting of a specific concept. That's why I included all the CSS code

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 13 above)
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: This handles the updates of the mappings data, when I changed from the Async Library I changed the way the function(updateAsyncSelectValue) to handle mappings is called in my input  so I had to write a test for it.

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 14]
```diff
@@ -86,4 +87,183 @@ describe('Test suite for dictionary concepts components', () => {
...
+    const inputField = wrapper.find('CreateMapping');
+    const spy = jest.spyOn(inputField.instance(), 'handleInputChange');
+    inputField.find('#searchInputCiel').simulate('change');
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('should handle change when a user inputs concept name mappings data for new rows when isNew is true', () => {
+    const newProps = {
+      ...props,
+      source: 'CIEL',
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Don't we have a constant for this?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 15]
```diff
@@ -86,4 +87,183 @@ describe('Test suite for dictionary concepts components', () => {
...
+    const inputField = wrapper.find('CreateMapping');
+    const spy = jest.spyOn(inputField.instance(), 'handleInputChange');
+    inputField.find('#searchInputCielIsnew').simulate('change');
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('Should handle key down event when a user presses the enter button to search mappings for existing rows', () => {
+    const newProps = {
+      ...props,
+      source: 'CIEL',
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Don't we have a constant for this?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 16]
```diff
@@ -86,4 +87,183 @@ describe('Test suite for dictionary concepts components', () => {
...
+    const spy = jest.spyOn(inputField.instance(), 'handleKeyPress');
+    const event = { key: 'Space' };
+    inputField.find('#searchInputCiel').simulate('keyDown', event);
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('Should handle click event when a user click to select a prefered concept name for existing rows', () => {
+    const newProps = {
+      ...props,
+      source: 'CIEL',
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Don't we have a constant for this?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 14 above)
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated it to use the default source constant `INTERNAL_MAPPING_DEFAULT_SOURCE`

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 15 above)
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated it to use the default source constant `INTERNAL_MAPPING_DEFAULT_SOURCE`

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 16 above)
Comment (judeinno) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated it to use the default source constant `INTERNAL_MAPPING_DEFAULT_SOURCE`
